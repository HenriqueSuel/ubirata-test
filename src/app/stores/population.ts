import { create } from "zustand";
import { devtools } from "zustand/middleware";

import {
  FeatureCollection,
  PopulationNeighborhoods,
  Properties,
} from "../interfaces/neighborhoodGeometries.interface";
import { getApi } from "../service/api.service";
import { handleSelectFilter } from "../utils/populations.utils";

interface IHandleSelectProps {
  years: string;
  idNeighborhood: number;
}

export interface PopulationState {
  selectedData: {
    years: string;
    idNeighborhood: number | null;
    neighborhood: Properties | null;
    population: PopulationNeighborhoods[];
  };
  filters: {
    neighborhoods: Properties[];
    years: string[];
  };
  neighborhoodGeometries: FeatureCollection | null;
  populationNeighborhoods: PopulationNeighborhoods[];
  handleSelectData: ({
    years,
    idNeighborhood,
  }: Partial<IHandleSelectProps>) => void;
  handleGetData: () => Promise<void>;
}

const usePopulation = create<PopulationState>()(
  devtools((set, get) => ({
    selectedData: {
      years: "",
      idNeighborhood: null,
      neighborhood: null,
      population: [],
    },
    filters: {
      neighborhoods: [],
      years: [],
    },
    neighborhoodGeometries: null,
    populationNeighborhoods: [],
    handleGetData: async () => {
      const [populationNeighborhoods, neighborhoodGeometries] =
        await Promise.all([
          getApi("mocks/populacao_bairros.json") as Promise<
            PopulationNeighborhoods[]
          >,
          getApi("mocks/geometrias_bairros.json") as Promise<FeatureCollection>,
        ]);

      const filters = handleSelectFilter(
        neighborhoodGeometries,
        populationNeighborhoods
      );

      const populationInitial = populationNeighborhoods.filter(
        (population) => population.id_geometria === filters.neighborhoods[0].id
      );

      set(() => ({
        filters,
        selectedData: {
          idNeighborhood: filters.neighborhoods[0].id,
          neighborhood: filters.neighborhoods[0],
          population: populationInitial,
          years: filters.years[0],
        },
        neighborhoodGeometries,
        populationNeighborhoods,
      }));
    },
    handleSelectData: ({ idNeighborhood, years }) => {
      const existingIdNeighborhood =
        idNeighborhood || get().selectedData.idNeighborhood;

      const neighborhood = get().filters.neighborhoods.find(
        (neighborhood) => neighborhood.id === existingIdNeighborhood
      );

      const populationNeighborhoods = get().populationNeighborhoods.filter(
        (item) => item.id_geometria === existingIdNeighborhood
      );

      set((state) => ({
        selectedData: {
          neighborhood: neighborhood || null,
          population: populationNeighborhoods,
          years: years || state.selectedData.years,
          idNeighborhood: existingIdNeighborhood,
        },
      }));
    },
  }))
);

export { usePopulation };
