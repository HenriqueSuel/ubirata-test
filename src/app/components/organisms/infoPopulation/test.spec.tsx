import { render, screen } from '@testing-library/react';
import { Mock } from 'vitest';
import { usePopulation } from '../../../stores/population';
import { InfoPopulation } from './';

vi.mock('../../../stores/population', () => ({
    usePopulation: vi.fn(),
}));

const mockNeighborhood = { id: 1, name: 'Neighborhood 1' };
const mockPopulation = [
    { ano: '2021', populacao: 10000 },
    { ano: '2022', populacao: 8000 },
];

describe('InfoPopulation', () => {

    beforeEach(() => {
        (usePopulation as unknown as Mock).mockReturnValue({
            selectedData: { neighborhood: mockNeighborhood, population: mockPopulation },
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it('should render InfoPopulation', () => {
        const { container } = render(<InfoPopulation />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render neighborhood name correctly', () => {
        render(<InfoPopulation />);
        const neighborhoodName = screen.getByText(mockNeighborhood.name);
        expect(neighborhoodName).toBeInTheDocument();
    });

    it('should render population data correctly', () => {
        render(<InfoPopulation />);
        mockPopulation.forEach(({ ano, populacao }) => {
            const populationData = screen.getByText(`Ano ${ano}: ${populacao}`);
            expect(populationData).toBeInTheDocument();
        });
    });
});
