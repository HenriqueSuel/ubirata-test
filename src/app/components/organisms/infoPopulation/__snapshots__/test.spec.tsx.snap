// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfoPopulation > should render InfoPopulation 1`] = `
<div
  class="max-w-sm rounded overflow-hidden shadow-lg absolute top-10 -right-1"
>
  <div
    class="px-4 py-4 bg-black"
  >
    <p
      class="text-xl pb-4"
    >
      Neighborhood 1
    </p>
    <p
      class="text-sm"
    >
      Ano 
      2021
      : 
      10000
    </p>
    <p
      class="text-sm"
    >
      Ano 
      2022
      : 
      8000
    </p>
  </div>
</div>
`;
